calico-vpp-agent/watchers/net_watcher.go

```diff
@@ -0,0 +1,210 @@
+	vpp                *vpplink.VppLink
+	client             client.WithWatch
+	stop               chan struct{}
+	networkDefinitions map[string]*NetworkDefinition
+}
+
+func NewNetWatcher(vpp *vpplink.VppLink, log *logrus.Entry) *NetWatcher {
+	kubernetesClient, err := NewClient(10 * time.Second, []func(s *runtime.Scheme) error{networkv3.AddToScheme})
+	if err != nil {
+		panic(fmt.Errorf("failed instantiating kubernetes client: %v", err))
+	}
+	w := NetWatcher{
+		log:                log,
+		vpp:                vpp,
+		client:             *kubernetesClient,
+		stop:               make(chan struct{}),
+		networkDefinitions: make(map[string]*NetworkDefinition),
+	}
+	return &w
+}
+
+func (w *NetWatcher) WatchNetworks(t *tomb.Tomb) error {
+	w.log.Infof("Net watcher starts")
+	netList := &networkv3.NetworkList{}
+	err := w.client.List(context.Background(), netList, &client.ListOptions{})
+	if err != nil {
+		return err
+	}
+	networks := make(map[string]*NetworkDefinition)
+	for _, net := range netList.Items {
+		netDef, err := w.OnNetAdded(&net)
+		if err != nil {
+			w.log.Error(err)
+			return err
+		}
+		networks[netDef.Name] = netDef
+	}
+	common.SendEvent(common.CalicoVppEvent{
+		Type: common.NetsSynced,
+		New:  networks,
+	})
+
+restart:
+	netList = &networkv3.NetworkList{}
+	watcher, err := w.client.Watch(context.Background(), netList, &client.ListOptions{})
+	if err != nil {
+		w.log.Errorf("couldn't watch pods: %s", err)
+	}
+	w.watching(watcher)
+	goto restart
+}
+
+func (w *NetWatcher) watching(watcher watch.Interface) bool {
+	for {
+		update, ok := <-watcher.ResultChan()
+		if !ok {
+			w.log.Warn("network watch channel closed")
+			return true
+		}
+		switch update.Type {
+		case watch.Added:
+			net := update.Object.(*networkv3.Network)
+			netDef, err := w.OnNetAdded(net)
+			if err != nil {
+				w.log.Error(err)
+			}
+			common.SendEvent(common.CalicoVppEvent{
+				Type: common.NetAdded,
+				New:  netDef,
+			})
+		case watch.Deleted:
+			oldNet := update.Object.(*networkv3.Network)
+			w.OnNetDeleted(oldNet.Name)
+		case watch.Modified:
+			w.log.Warn("network changed, not yet supported!")
+		}
+	}
+}
+
+func (w *NetWatcher) Stop() {
+	close(w.stop)
+}
+
+func (w *NetWatcher) OnNetAdded(net *networkv3.Network) (*NetworkDefinition, error) {
+	netDef, err := w.CreateVRFsForNet(net.Name, uint32(net.Spec.VNI), net.Spec.Range)
+	if err != nil {
+		return nil, err
+	}
+	return netDef, nil
+}
+
+func (w *NetWatcher) OnNetChanged(old, new *networkv3.Network) {
+	// TODO handle network change
+}
+
+func (w *NetWatcher) OnNetDeleted(netName string) error {
+	w.log.Infof("deleting network %s", netName)
+	netDef, err := w.DeleteNetVRFs(netName)
+	if err != nil {
+		return err
+	}
+	common.SendEvent(common.CalicoVppEvent{
+		Type: common.NetDeleted,
+		Old:  netDef,
+	})
+	return nil
+}
+
+func getNetworkVrfName(networkName string, suffix string) string {
+	return fmt.Sprintf("pod-%s-table-%s", networkName, suffix)
```

Comment:
Just as a note, the name is capped at 64 characters, but I guess this should be enough (I don't know what network names typically are). Otherwise, it's probably worth doing something like  `GetVrfTag`

Reply:
I guess it is a low probability that this exceeds 64 char, we can keep it for now